idx/memory/priority_lock.go

```diff
@@ -0,0 +1,124 @@
+package memory
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// PriorityRWMutex is a mutex that allows fair access to high priority read operations while
+// blocking low priority operations if a write lock is waiting.
+type PriorityRWMutex struct {
+	lock        sync.RWMutex
+	lowPrioLock sync.RWMutex
+}
+
+type BlockContext struct {
+	lock         *PriorityRWMutex
+	preLockTime  time.Time
+	postLockTime time.Time
+	postOpTime   time.Time
+}
+
+// RLockHigh allows *guaranteed fast* (or very high priority) ops to acquire the read lock as
+// quickly as possible. Note: This can block write lock acquisition
+func (pm *PriorityRWMutex) RLockHigh() {
+	// Don't increment activeReads just grab the lock. If an active write is holding the lock it
+	// will block, otherwise it *may* block a write acquisition, but that is opted into.
+	pm.lock.RLock()
+}
+
+// RUnlockHigh unlocks read lock called via RLockHigh
+func (pm *PriorityRWMutex) RUnlockHigh() {
+	pm.lock.RUnlock()
+}
+
+// Synonym for RLockLow
+func (pm *PriorityRWMutex) RLock() BlockContext {
+	return pm.RLockLow()
+}
+
+// Synonym for RUnlockLow
+func (pm *PriorityRWMutex) RUnlock(bc *BlockContext) {
+	pm.RUnlockLow(bc)
+}
+
+// RLockLow will block until requested writes complete. New writes that come in also get priority.
+func (pm *PriorityRWMutex) RLockLow() BlockContext {
+	// Wait for any pending writes
+	bc := BlockContext{lock: pm, preLockTime: time.Now()}
+	pm.lowPrioLock.RLock()
+	pm.lock.RLock()
```

Comment:
I'm not entirely convinced we actually need to lock both of these here, if `Lock` is guaranteed to acquire both write locks.